### PR TITLE
changefeedccl: smart family selection with expressions

### LINKED ISSUE #82607: changefeedccl: Smart family selection with expressions
Changefeeds return an error when running predicate changefeed against "each" family.
This restriction is unnecessary.  If the expression does not reference any columns outside
of primary keys, there is nothing that should prevent us from running predicate changefeed.
For example `changefeed AS select * where NOT cdc_is_delete()` is absolutely fine even when 
running against each family.

In addition, we don't have to rely on the user to specify which column family they are interested
in.  This can be determined based on the expressions.
The overall idea is: walk the expressions, collecting the set of column families referenced by all expressions.
3 cases to consider:
  * 0 families found: i.e. expression simply did not reference any named column; for example
    `CHANGEFEED .. AS SELECT * WHERE NOT cdc_is_delete()`.
    In this case, the expression is fine, regardless of whether or not we have multiple families
  * 1 family -- great, fetch that family
  *  more than 1 family -- an error.  
  
Jira issue: CRDB-16535

### PR DESCRIPTION
**Pull request by surajr10**
ccl: implemented functionality to determine the number of column families that
are referenced by a SELECT statement in changefeed expressions and handle
appropriately

Resolves: https://github.com/cockroachdb/cockroach/issues/82607

Release note (enterprise change): implemented functionality to determine the
number of column families that are referenced by a SELECT statement in
changefeed expressions and handle appropriately

### CODE REVIEW

**Review by HonoreDB — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84764)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @surajr10)

___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7az3kM-WGbY4DRuUCh:-N7az3kM-WGbY4DRuUCi:b-b5wz92) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/ccl/changefeedccl/cdceval/validation.go#L132)):*
> ```Go
> 	allFamilies := desc.GetFamilies()
> 
> 	if len(allFamilies) == 1 && allFamilies[0].Name == "primary" {
> ```

Despite the type name, I don't think we need the Name == "primary" check. We want the behavior to be the same if there's only one family but it was named something else, (e.g. `CREATE TABLE onefamwithcustomname(id int primary key FAMILY f1)`).
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7b-1MHEKx4xXG6jKcQ:-N7b-1MHEKx4xXG6jKcR:b-fwwrxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/ccl/changefeedccl/cdceval/validation.go#L135)):*
> ```Go
> 		publicColSet := catalog.MakeTableColSet(desc.PublicColumnIDs()...)
> 		colSet := catalog.MakeTableColSet(allFamilies[0].ColumnIDs...)
> 		if colSet.Equals(publicColSet) {
> ```

I don't follow what this comparison is doing. It looks like you're validating that if there's only one family, that family contains all public columns for the table? But I think we can count on that. You can probably just blindly set the type to PRIMARY_FAMILY_ONLY if there's only one table.
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7b1fN00TH82kjiqy-X:-N7b1fN00TH82kjiqy-Y:b-3jjfhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/ccl/changefeedccl/cdceval/validation.go#L151)):*
> ```Go
> 		colSet := catalog.MakeTableColSet(family.ColumnIDs...)
> 		for colID := range cv.columns {
> 			if colSet.Contains(colID) {
> ```

Nit: ColSet is implemented using FastIntSet, which has an Intersects() method. If you used that to add an Intersects() method to https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/catalog/table_col_set.go, and made cv.columns also be a TableColSet, this method could be one line shorter and probably a little faster.

Edit: Looks like you're already adding Equals. Maybe replace that with Intersects as I don't think you actually need Equals.
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7b3TTk1oZZVgJRDdkP:-N7b3TTk1oZZVgJRDdkQ:b-nk97bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/ccl/changefeedccl/cdceval/validation.go#L159)):*
> ```Go
> 
> 	if len(referencedFamilies) > 1 {
> 		target.Type = jobspb.ChangefeedTargetSpecification_EACH_FAMILY
> ```

Haven't read through the whole PR yet but if we get to this branch we're going to need to error out, maybe let this method return an error or at least add a comment here that this is an invalid state?
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 410 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7b5PTP4Jo__xAbwQ-v:-N7b5PTP4Jo__xAbwQ-w:b-acd4m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/ccl/changefeedccl/cdceval/validation.go#L410)):*
> ```Go
> 
> 	case *tree.ColumnItem:
> 		col, err := c.desc.FindColumnWithName(e.ColumnName)
> ```

I wonder if there's a way to get a working `tree.ColumnItemResolver` here. @miretskiy is probably looking into the same question as part of his reimplementing cdc_prev. It'd be nice to be able to outsource worrying about things like capitalization and name prefixes to existing code.
___
*[`pkg/sql/catalog/table_col_set.go` line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7b713d5BMAugKWGYZt:-N7b713eCf8maUMCj-Az:b6g302y) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/sql/catalog/table_col_set.go#L92)):*
> ```Go
> func (s TableColSet) String() string { return s.set.String() }
> 
> // Equals returns whether or not s and other contain the same elements
> ```

nit: period at the end of the comment. Or I guess remove this method if you don't end up using it.
___
*[`pkg/ccl/changefeedccl/cdceval/validation_test.go` line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7b7FQuEHWUnt4W38Fl:-N7b7FQuEHWUnt4W38Fm:b-y1k3d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/ccl/changefeedccl/cdceval/validation_test.go#L97)):*
> ```Go
> 			desc:      bazDesc,
> 			stmt:      "SELECT a, b, c FROM baz",
> 			expectErr: `projections and filter cannot be used when running against multifamily table`,
> ```

This error message should change, something like `expressions can't reference columns from more than one column family`. Or `expressions can't reference non-primary-key columns from more than one column family` if that's what we decided, I forget.
___
*[`pkg/ccl/changefeedccl/cdceval/validation_test.go` line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7b7yw38i1hD5O3TT9A:-N7b7yw38i1hD5O3TT9B:brc8uze) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/ccl/changefeedccl/cdceval/validation_test.go#L103)):*
> ```Go
> 			desc:      bazDesc,
> 			stmt:      "SELECT * FROM baz",
> 			expectErr: `projections and filter cannot be used when running against multifamily table`,
> ```

In this case I think the error (and behavior) we want is the same as a regular changefeed: `CHANGEFEED targeting a table (hasfams) with multiple column families requires WITH split_column_families and will emit multiple events per row.`
___
*[`pkg/ccl/changefeedccl/cdceval/validation_test.go` line 145 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7b8i5a3BHwIZ5dYGuO:-N7b8i5a3BHwIZ5dYGuP:bmd2esk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/ccl/changefeedccl/cdceval/validation_test.go#L145)):*
> ```Go
> 			stmt:       "SELECT a, b FROM baz",
> 			expectStmt: fmt.Sprintf("SELECT a, b FROM [%d AS baz]", bazDesc.GetID()),
> 		},
> ```

Other interesting cases: 

`SELECT a,c FROM baz` Counterintuitively I think this should succeed.

`SELECT b, b+1 AS c FROM baz` Should succeed, the validator can't get confused by the alias.

`SELECT b, c as a FROM baz` Should error,  the validator can't get confused by the alias.

`SELECT B FROM BAZ`. Should succeed, the validator can't sulk if you shout at it.

`SELECT baz.b, baz.public.b FROM baz` Ideally should succeed but I don't think we we support this on master yet? If not disregard.

`SELECT b FROM baz WHERE c IS NULL` Should error.

`SELECT b, substring(c,1,2) FROM baz` Should error.

`SELECT b = 'c' FROM baz` Should succeed.

**Comment by miretskiy** [comment]
fyi: have you seen `rowenc.NeededColumnFamilyIDs` utility function? might come in handy

**Review by HonoreDB — APPROVED** [review]
> Reviewed 3 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84764)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @surajr10)

___
*[`pkg/ccl/changefeedccl/changefeed_stmt.go` line 383 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N8-0TLx3F3Wy5VBJmcK:-N8-0TLx3F3Wy5VBJmcL:b-uyovqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/58fcbbe19fa62197ae51b8bbff0a16b658198f9b/pkg/ccl/changefeedccl/changefeed_stmt.go#L383)):*
> ```Go
> 		// Serialize changefeed expression.
> 		normalized, _, err := validateAndNormalizeChangefeedExpression(
> 			ctx, p, changefeedStmt.Select, targetDescs, targets, opts.IncludeVirtual(), opts.IsSet(changefeedbase.OptSplitColumnFamilies),
> ```

Nit: You can replace `opts.IncludeVirtual(), opts.IsSet(changefeedbase.OptSplitColumnFamilies)` with `opts.GetCanHandle()`.
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 410 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N7b5PTP4Jo__xAbwQ-v:-N8-2IyLEf_McKQ0przV:bf0mapv) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc1c3f53462b4502024bd59c6fb670ddd657d75a/pkg/ccl/changefeedccl/cdceval/validation.go#L410)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I wonder if there's a way to get a working `tree.ColumnItemResolver` here. @miretskiy is probably looking into the same question as part of his reimplementing cdc_prev. It'd be nice to be able to outsource worrying about things like capitalization and name prefixes to existing code.
</blockquote></details>

Will defer to @miretskiy on this but given all the test cases you added I feel good about this implementation.
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 150 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N8-1j--DRHnXY-NdR3m:-N8-1j--DRHnXY-NdR3n:br5f5pe) ([raw file](https://github.com/cockroachdb/cockroach/blob/58fcbbe19fa62197ae51b8bbff0a16b658198f9b/pkg/ccl/changefeedccl/cdceval/validation.go#L150)):*
> ```Go
> 	}
> 
> 	if nonKeyColSet.Len() == 0 {
> ```

Please add a comment about what this case is doing, it's correct but a bit confusing.
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 427 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-N8-2WU8ACjmE1NraOtH:-N8-2WU8ACjmE1NraOtI:brgpumw) ([raw file](https://github.com/cockroachdb/cockroach/blob/58fcbbe19fa62197ae51b8bbff0a16b658198f9b/pkg/ccl/changefeedccl/cdceval/validation.go#L427)):*
> ```Go
> 			return false, expr
> 		}
> 		colIdx := col.GetID()
> ```

Nit: colID, not colIdx. Not always the same thing.

**Review by HonoreDB — COMMENTED** [review]
Looks great to me! Approved, pending some cosmetic comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84764#-:-N8-2w-fE-UgPp9IjD3Z:bgcqgh5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @surajr10)

**Pull request merged**
